onnxscript/rewriter/pattern.py

```diff
@@ -927,57 +1023,38 @@ def _apply_deltas(
-                    graph_or_function.values[output_name] = value
-            path_2 = True
-
-    assert not to_delete or not path_2, (
-        "Two different rules were applied. It will solved later. "
-        "Right now, the functions assumes all the changes come from one "
-        "rule."
-    )
-
-    if path_2:
-        for _, delta in deltas:
-            deleted_nodes, inserted_nodes = delta
-            inserted_input_output = []
-            for nd in inserted_nodes:
-                inserted_input_output += nd.inputs + nd.outputs
-            for old_node in deleted_nodes[0:-1]:
-                # Delete intermediary outputs from graph that are not used as
-                # outputs of the graph
-                for output in old_node.outputs:
-                    if not output.is_output and output not in inserted_input_output:
-                        graph_or_function.values.pop(output.name)
-                nodes.remove(old_node)
-
-    for i in to_delete:
-        position = existing_ids[i][0]
-        nodes[position] = None
-
-    for position, insert in sorted(to_insert.items(), reverse=True):
-        for v in reversed(insert):
-            nodes.insert(position, v)
-
-    position_to_delete = []
-    for i, n in enumerate(nodes):
-        if n is None:
-            position_to_delete.append(i)
-
-    for p in reversed(position_to_delete):
-        del nodes[p]
+            # Reconnect the users of the deleted node to use the new outputs
+            for last_deleted_output, last_inserted_output in zip(
+                last_deleted.outputs, last_inserted.outputs
+            ):
+                for node, index in last_deleted_output.consumers():
+                    node.replace_input_with(index, last_inserted_output)
+
+                # Update graph/function outputs if the node generates output
+                for old_output, new_output in zip(last_deleted.outputs, last_inserted.outputs):
+                    for idx, graph_or_function_output in enumerate(graph_or_function.outputs):
+                        if graph_or_function_output is old_output:
+                            graph_or_function.outputs[idx] = new_output
+
+            # insert new nodes after the index node
+            # TODO(justinchuby): Do not access by index [i]
+            graph_or_function.insert_after(graph_or_function.nodes[i], inserted_nodes)
+
+            for old_node in deleted_nodes:
+                graph_or_function.remove(old_node)
+
+    for replaced_node, inserted_nodes in to_insert:
+        graph_or_function.insert_after(replaced_node, inserted_nodes)
+        # TODO: improve this
+        # This is updating the graph/function outputs to use the new outputs
+        for inserted_node in inserted_nodes:
+            for new_output in inserted_node.outputs:
+                if (index := new_output.meta.get(_ir_utils.GRAPH_OUTPUT_META_KEY)) is not None:
```

Comment:
What is this for? Feeling a bit lost here.

Reply:
Here we had the issue of we don't know which matched pattern node should be an output to the graph with how the nodes are passed in to this function right now. So we use a meta field to store the info and update the graph output as needed.

Alternatively, we can pass a mapping of {old_value: new_value} to simply call `old_value.is_graph_output()` when updating the connections.

Reply:
I see. Ok, I guess we can deal with this in the next phase/PR for simplification (of the API for this function).